build: Optionally enable -Wzero-as-null-pointer-constant

Empact opened the pull request:
~~**Note:** This is based on #14920~~

This enables `-Wzero-as-null-pointer-constant`, while avoiding applying it to dependencies.

~~Used [Qt::NoItemFlags](http://doc.qt.io/qt-5/qt.html#ItemFlag-enum),  [Qt::Widget](http://doc.qt.io/qt-5/qt.html#WindowType-enum) instead where applicable.~~

See #10483, #10645, #13802 for more on `nullptr`.

practicalswift committed:
Don't use zero as null pointer constant (-Wzero-as-null-pointer-constant)

Empact committed:
build: Enable -Wzero-as-null-pointer-constant under suppress_external_warnings

And -Werror=zero-as-null-pointer-constant if werror are enabled.

While disabling it in the following dependencies:
* LevelDb via build flags
* Tinyformat via pragmas due to external library
* Sqlite via pragmas due to use of SQLITE_STATIC
* util_tests via pragmas due to use of SIG_DFL

practicalswift commented:
Concept ACK and in accordance with our developer notes.

Using zero as null pointer constant is sloppy at best and dangerous at worst :-)

Empact commented:
This build is timing out, seeing if I can do anything about that:
https://travis-ci.org/bitcoin/bitcoin/jobs/475830791

MarcoFalke commented:
Could submit the src/qt changes as separate pull request, since those are the bulk of the changes?

practicalswift commented:
@MarcoFalke Perhaps 10f81e916d5fbb2751d03d06901f68d9b507807c (including the two non QT fixes) can be posted as a separate PR since they are all trivial to review?

practicalswift commented:
@Empact What about `-Werror=zero-as-null-pointer-constant` as discussed in https://github.com/bitcoin/bitcoin/pull/15114#issuecomment-451953456?

Empact commented:
Enabled now. 👍

practicalswift commented:
utACK 873145a56315ba032aa100646d682ec31d39e8d9

Empact commented:
Rebased for #13216

Empact commented:
Rebased for #15154

Empact commented:
AppVeyor failure is spurious, solved by #15171. Now that  #15114 is in, the diff is much smaller here. Still should start with #14920.

practicalswift commented:
@Empact Appveyor restarted.

Empact commented:
Rebased for #15175

practicalswift commented:
utACK b4996079cda89711c2b8b80adbbb06a80f7b9c86

Empact commented:
Rebased due to conflicts in #15377

Empact commented:
Rebased for #15288

practicalswift commented:
FWIW: I've verified that a disassembly of the `bitcoind` binary built with this patch applied is identical to a disassembly of the `bitcoind` binary built against `master` (as expected).

practicalswift commented:
utACK 586a564f4c382bfe308749f573ca738e87a188da

laanwj commented:
I get tons of these warnings (almost on every file):
```
/home/user/src/bitcoin/src/tinyformat.h:119:32: warning: unknown warning group '-Wzero-as-null-pointer-constant', ignored [-Wunknown-pragmas]
#pragma GCC diagnostic ignored "-Wzero-as-null-pointer-constant"
                               ^
```
This is with `clang version 3.8.1-24 (tags/RELEASE_381/final)` (which is old as the hills but we apparently still support).

promag commented:
Failing in travis with
```
In file included from ./index/base.h:8:0,
                 from index/base.cpp:6:
./dbwrapper.h:16:10: fatal error: leveldb/db.h: No such file or directory
 #include <leveldb/db.h>
```

Empact commented:
Yeah @promag this is wip: rebasing and addressing @laanwj’s comment

practicalswift commented:
@Empact Seems like the current version does not compile?

kallewoof reviewed (approved):
ACK fc19862337f0764104a918f22a97c47fb5dcc304

* Reviewed code
* Tested locally; managed to trigger warning when swapping a `nullptr` for `0`. (Was unable to get a documentation warning by messing with a `@param` varname , though, for some reason...)

practicalswift commented:
I'll happily re-review and re-ACK after rebase -- hopefully we can get this in soon :)

laanwj commented:
Fails in leveldb (shouldn't it skip dependencies?):
```
./leveldb/db/skiplist.h:328:21: error: zero as null pointer constant [-Werror=zero-as-null-pointer-constant]
       head_(NewNode(0 /* any key will do */, kMaxHeight)),
                     ^
cc1plus: some warnings being treated as errors
```

hebasto commented on configure.ac:
```diff
@@ -442,6 +442,10 @@ if test "x$enable_werror" = "xyes"; then
 
   if test x$suppress_external_warnings != xno ; then
     AX_CHECK_COMPILE_FLAG([-Werror=documentation],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=documentation"],,[[$CXXFLAG_WERROR]])
+    AX_CHECK_COMPILE_FLAG([-Werror=zero-as-null-pointer-constant],[
+      WARN_CXXFLAGS="$WARN_CXXFLAGS -Werror=zero-as-null-pointer-constant"
+      AC_DEFINE(HAVE_ZERO_AS_NULL_POINTER_CONSTANT, 1, [Define this symbol if -Wzero-as-null-pointer-constant is available])
+    ],,[[$CXXFLAG_WERROR]])
```
One pair of `[]` is redundant:
```suggestion
    AX_CHECK_COMPILE_FLAG([-Werror=zero-as-null-pointer-constant], [
      WARN_CXXFLAGS="$WARN_CXXFLAGS -Werror=zero-as-null-pointer-constant"
      AC_DEFINE(HAVE_ZERO_AS_NULL_POINTER_CONSTANT, 1, [Define this symbol if -Wzero-as-null-pointer-constant is available])
    ], [], [$CXXFLAG_WERROR])
```

Empact replied:
I'm an autoconf novice - I take it we're not concerned about macro conflicts with the CXXFLAG_WERROR as its contents are of a different form than uppercase macro definitions?

> When you use the same text in a macro argument, you must therefore have an extra quotation level (since one is stripped away by the macro substitution). In general, then, it is a good idea to use double quoting for all literal string arguments, either around just the problematic portions, or over the entire argument:

>     AC_MSG_WARN([[AC_DC] stinks  --Iron Maiden])
>     AC_MSG_WARN([[AC_DC stinks  --Iron Maiden]])
https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.69/html_node/Autoconf-Language.html#Autoconf-Language

hebasto commented on configure.ac:
```diff
@@ -472,6 +476,10 @@ if test "x$CXXFLAGS_overridden" = "xno"; then
 
   if test x$suppress_external_warnings != xno ; then
     AX_CHECK_COMPILE_FLAG([-Wdocumentation],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wdocumentation"],,[[$CXXFLAG_WERROR]])
+    AX_CHECK_COMPILE_FLAG([-Wzero-as-null-pointer-constant],[
+      WARN_CXXFLAGS="$WARN_CXXFLAGS -Wzero-as-null-pointer-constant"
+      AC_DEFINE(HAVE_ZERO_AS_NULL_POINTER_CONSTANT, 1, [Define this symbol if -Wzero-as-null-pointer-constant is available])
+    ],,[[$CXXFLAG_WERROR]])
```
One pair of `[]` is redundant:
```suggestion
    AX_CHECK_COMPILE_FLAG([-Wzero-as-null-pointer-constant], [
      WARN_CXXFLAGS="$WARN_CXXFLAGS -Wzero-as-null-pointer-constant"
      AC_DEFINE(HAVE_ZERO_AS_NULL_POINTER_CONSTANT, 1, [Define this symbol if -Wzero-as-null-pointer-constant is available])
    ], [], [$CXXFLAG_WERROR])
```

hebasto reviewed (commented):
Approach ACK 2036b7606f16a76a300f067824168de6536c60c8

I think the following patch could help:
```diff
--- a/src/Makefile.leveldb.include
+++ b/src/Makefile.leveldb.include
@@ -36,7 +36,7 @@ LEVELDB_CPPFLAGS_INT += -DLEVELDB_PLATFORM_POSIX
 endif
 
 leveldb_libleveldb_a_CPPFLAGS = $(AM_CPPFLAGS) $(LEVELDB_CPPFLAGS_INT) $(LEVELDB_CPPFLAGS)
-leveldb_libleveldb_a_CXXFLAGS = $(filter-out -Wconditional-uninitialized -Werror=conditional-uninitialized -Wsuggest-override -Werror=suggest-override, $(AM_CXXFLAGS)) $(PIE_FLAGS)
+leveldb_libleveldb_a_CXXFLAGS = $(filter-out -Wconditional-uninitialized -Werror=conditional-uninitialized -Wsuggest-override -Werror=suggest-override -Wzero-as-null-pointer-constant -Werror=zero-as-null-pointer-constant, $(AM_CXXFLAGS)) $(PIE_FLAGS)
 
 leveldb_libleveldb_a_SOURCES=
 leveldb_libleveldb_a_SOURCES += leveldb/port/port_stdcxx.h
```

In the OP could drop "Used Qt::NoItemFlags, Qt::Widget instead where applicable." ?

In commit "Don't use zero as null pointer constant (-Wzero-as-null-pointer-constant)" (2036b7606f16a76a300f067824168de6536c60c8) could drop "Protobuf generated files via pragmas inserted via src/Makefile.am" ?

The name `HAVE_ZERO_AS_NULL_POINTER_CONSTANT` seems confusing, as actually it means an activated warning, not a feature.

Empact commented:
@hebasto thanks for the review - I switched to `HAVE_W_ZERO_AS_NULL_POINTER_CONSTANT` - do you think that's clearer? Open to suggestions.

I'm unclear on the `AX_CHECK_COMPILE_FLAG` feedback - here I'm following the example from other `AX_CHECK_COMPILE_FLAG`s in the same file, e.g.:

```
AX_CHECK_COMPILE_FLAG([-Wdocumentation],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wdocumentation"],,[[$CXXFLAG_WERROR]])
AX_CHECK_COMPILE_FLAG([-Wduplicated-cond],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wduplicated-cond"],,[[$CXXFLAG_WERROR]])
```

Why should this be different?

hebasto commented:
https://cirrus-ci.com/task/6169333320646656?logs=ci#L2845
```
test/util_tests.cpp:1805:48: error: zero as null pointer constant [-Werror,-Wzero-as-null-pointer-constant]
    void (*old_handler)(int) = signal(SIGCHLD, SIG_DFL);
                                               ^
/usr/include/x86_64-linux-gnu/bits/signum-generic.h:29:37: note: expanded from macro 'SIG_DFL'
#define SIG_DFL  ((__sighandler_t)  0)  /* Default action.  */
                                    ^
1 error generated.
```

practicalswift commented:
cr ACK 7f0f2ca49f23cfd9e28892137e13f057a2152a91: patch looks correct

hebasto commented on src/tinyformat.h:
```diff
@@ -124,6 +124,17 @@
+#if defined(HAVE_W_ZERO_AS_NULL_POINTER_CONSTANT)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wzero-as-null-pointer-constant"
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wzero-as-null-pointer-constant"
```
With gcc 9.3.0:
```
./tinyformat.h:134: warning: ignoring #pragma clang diagnostic [-Wunknown-pragmas]
  134 | #pragma clang diagnostic push
      | 
./tinyformat.h:135: warning: ignoring #pragma clang diagnostic [-Wunknown-pragmas]
  135 | #pragma clang diagnostic ignored "-Wzero-as-null-pointer-constant"
      | 
```

MarcoFalke replied:
Might be better to do with clang-tidy or so externally instead of maiming the source code?

fanquake replied:
> Might be better to do with clang-tidy

Have done this in #24971.

hebasto commented on src/tinyformat.h:
```diff
@@ -1163,4 +1174,9 @@ std::string format(const std::string &fmt, const Args&... args)
 
+#if defined(HAVE_W_ZERO_AS_NULL_POINTER_CONSTANT)
+#pragma GCC diagnostic pop
+#pragma clang diagnostic pop
```
With gcc 9.3.0:
```
./tinyformat.h:1179: warning: ignoring #pragma clang diagnostic [-Wunknown-pragmas]
 1179 | #pragma clang diagnostic pop
      | 
```

hebasto commented on src/wallet/sqlite.cpp:
```diff
@@ -365,6 +365,14 @@ void SQLiteBatch::Close()
     }
 }
 
+// disable -Wzero-as-null-pointer-constant due to SQLITE_STATIC
```
If I remove changes in this file I won't receive warning with both gcc 9.3.0 and clang 10.0.0.
What cases are guarded here?

Having system sqlite 3.31.1.

Empact committed:
Ignore gcc -Wunknown-pragmas in tinyformat.h

As without this, GCC warns on the following unrecognized clang pragmas.
https://github.com/bitcoin/bitcoin/pull/15112#discussion_r623251609

Empact commented:
Rebase for the move of `SQLITE_STATIC` that occurred in #23732

fanquake commented:
Going to close this in favour of #24971; where we outsource this check to external tooling, and don't have to cover the codebase in pragmas to satisfy compilers.

Empact commented:
Sounds good, one of these days I'll familiarize myself with clang-tidy.
